<a href="https://colab.research.google.com/github/ZachLondo/Sales-Predictions/blob/main/ProjectOne_PartOne.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [36]:
filename = '/content/drive/MyDrive/01 Week 1: Python/sales_predictions.csv'
df = pd.read_csv(filename)
df.head(30)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0            FDA15        9.300          Low Fat         0.016047   
1            DRC01        5.920          Regular         0.019278   
2            FDN15       17.500          Low Fat         0.016760   
3            FDX07       19.200          Regular         0.000000   
4            NCD19        8.930          Low Fat         0.000000   
5            FDP36       10.395          Regular         0.000000   
6            FDO10       13.650          Regular         0.012741   
7            FDP10          NaN          Low Fat         0.127470   
8            FDH17       16.200          Regular         0.016687   
9            FDU28       19.200          Regular         0.094450   
10           FDY07       11.800          Low Fat         0.000000   
11           FDA03       18.500          Regular         0.045464   
12           FDX32       15.100          Regular         0.100014   
13           FDS46       17.600          Regular         0.047257   
14           FDF32       16.350          Low Fat         0.068024   
15           FDP49        9.000          Regular         0.069089   
16           NCB42       11.800          Low Fat         0.008596   
17           FDP49        9.000          Regular         0.069196   
18           DRI11          NaN          Low Fat         0.034238   
19           FDU02       13.350          Low Fat         0.102492   
20           FDN22       18.850          Regular         0.138190   
21           FDW12          NaN          Regular         0.035400   
22           NCB30       14.600          Low Fat         0.025698   
23           FDC37          NaN          Low Fat         0.057557   
24           FDR28       13.850          Regular         0.025896   
25           NCD06       13.000          Low Fat         0.099887   
26           FDV10        7.645          Regular         0.066693   
27           DRJ59       11.650          low fat         0.019356   
28           FDE51        5.925          Regular         0.161467   
29           FDC14          NaN          Regular         0.072222   

                Item_Type  Item_MRP Outlet_Identifier  \
0                   Dairy  249.8092            OUT049   
1             Soft Drinks   48.2692            OUT018   
2                    Meat  141.6180            OUT049   
3   Fruits and Vegetables  182.0950            OUT010   
4               Household   53.8614            OUT013   
5            Baking Goods   51.4008            OUT018   
6             Snack Foods   57.6588            OUT013   
7             Snack Foods  107.7622            OUT027   
8            Frozen Foods   96.9726            OUT045   
9            Frozen Foods  187.8214            OUT017   
10  Fruits and Vegetables   45.5402            OUT049   
11                  Dairy  144.1102            OUT046   
12  Fruits and Vegetables  145.4786            OUT049   
13            Snack Foods  119.6782            OUT046   
14  Fruits and Vegetables  196.4426            OUT013   
15              Breakfast   56.3614            OUT046   
16     Health and Hygiene  115.3492            OUT018   
17              Breakfast   54.3614            OUT049   
18            Hard Drinks  113.2834            OUT027   
19                  Dairy  230.5352            OUT035   
20            Snack Foods  250.8724            OUT013   
21           Baking Goods  144.5444            OUT027   
22              Household  196.5084            OUT035   
23           Baking Goods  107.6938            OUT019   
24           Frozen Foods  165.0210            OUT046   
25              Household   45.9060            OUT017   
26            Snack Foods   42.3112            OUT035   
27            Hard Drinks   39.1164            OUT013   
28                  Dairy   45.5086            OUT010   
29                 Canned   43.6454            OUT019   

    Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                        1999

1) How many rows and columns?

In [7]:
df.shape

(8523, 12)

8523 Rows, 12 Columns

2) What are the datatypes of each variable?

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


3) Are there duplicates? If so, drop any duplicates.

In [18]:
df.duplicated().sum()

0

4) Identify missing values.

In [19]:
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

5) Decide on how to address the missing values and do it! (This requires your judgement, so explain your choice).

In [22]:
df.isna()

Item_Identifier  Item_Weight  Item_Fat_Content  Item_Visibility  \
0               False        False             False            False   
1               False        False             False            False   
2               False        False             False            False   
3               False        False             False            False   
4               False        False             False            False   
...               ...          ...               ...              ...   
8518            False        False             False            False   
8519            False        False             False            False   
8520            False        False             False            False   
8521            False        False             False            False   
8522            False        False             False            False   

      Item_Type  Item_MRP  Outlet_Identifier  Outlet_Establishment_Year  \
0         False     False              False                      False   
1         False     False              False                      False   
2         False     False              False                      False   
3         False     False              False                      False   
4         False     False              False                      False   
...         ...       ...                ...                        ...   
8518      False     False              False                      False   
8519      False     False              False                      False   
8520      False     False              False                      False   
8521      False     False              False                      False   
8522      False     False              False                      False   

      Outlet_Size  Outlet_Location_Type  Outlet_Type  Item_Outlet_Sales  
0           False                 False        False              False  
1           False                 False        False              False  
2           False                 False        False              False  
3            True                 False        False              False  
4           False                 False        False              False  
...           ...                   ...          ...                ...  
8518        False                 False        False              False  
8519         True                 False        False              False  
8520        False                 False        False              False  
8521        False                 False        False              False  
8522        False                 False        False              False  

[8523 rows x 12 columns]

There is too much missing data to drop the rows completely. 

In [24]:
df['Item_Weight'].describe()

count    7060.000000
mean       12.857645
std         4.643456
min         4.555000
25%         8.773750
50%        12.600000
75%        16.850000
max        21.350000
Name: Item_Weight, dtype: float64

Based on the above description, I will replace the missing data with the average. This average looks to be the best fit because the item weights do not vary significantly, therefore, I believe the average will not be far off from the correct weight.

In [40]:
average_weight = df['Item_Weight'].mean()
df['Item_Weight'].fillna(average_weight, inplace=True)
df.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

For "Outlet size," there does not seem to be any correlation to the Identifier, Outlet type, Location Type, or Sales Volume. It is unclear whether or not this column will be relevant in predicting sales in the future, so I am hesitant to drop it altogether. Based on the mode only accounting for less than half of the collected data (below), I am also hesitant to replace this data with either the Median or Mode. I believe creating a new category will offer the best solution as it will allow me to retain the data without skewing a potentially useful column with faulty inputs.

In [30]:
df['Outlet_Size'].describe()

count       6113
unique         3
top       Medium
freq        2793
Name: Outlet_Size, dtype: object

In [37]:
df['Outlet_Size'].fillna('Missing', inplace=True)

In [38]:
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998     Missing               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

6) Confirm that there are no missing values after addressing them.

In [41]:
df.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

7) Find and fix any inconsistent categories of data (example: fix cat, Cat, and cats so that they are consistent) 

In [45]:
df['Item_Identifier'].value_counts()

FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64

n/a

In [47]:
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [48]:
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('LF','Low Fat')
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('reg','Regular')
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('low fat','Low Fat')
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

Nice

In [49]:
df['Item_Type'].value_counts()

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

n/a

In [50]:
df['Outlet_Size'].value_counts()

Medium     2793
Missing    2410
Small      2388
High        932
Name: Outlet_Size, dtype: int64

8) For any numerical columns, obtain the summary statistics of each (min, max, mean)

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                8523 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                8523 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [52]:
df['Item_Weight'].describe()

count    8523.000000
mean       12.857645
std         4.226124
min         4.555000
25%         9.310000
50%        12.857645
75%        16.000000
max        21.350000
Name: Item_Weight, dtype: float64

In [53]:
df['Item_Visibility'].describe()

count    8523.000000
mean        0.066132
std         0.051598
min         0.000000
25%         0.026989
50%         0.053931
75%         0.094585
max         0.328391
Name: Item_Visibility, dtype: float64

In [54]:
df['Item_MRP'].describe()

count    8523.000000
mean      140.992782
std        62.275067
min        31.290000
25%        93.826500
50%       143.012800
75%       185.643700
max       266.888400
Name: Item_MRP, dtype: float64

In [55]:
df['Outlet_Establishment_Year'].describe()

count    8523.000000
mean     1997.831867
std         8.371760
min      1985.000000
25%      1987.000000
50%      1999.000000
75%      2004.000000
max      2009.000000
Name: Outlet_Establishment_Year, dtype: float64

In [56]:
df['Item_Outlet_Sales'].describe()

count     8523.000000
mean      2181.288914
std       1706.499616
min         33.290000
25%        834.247400
50%       1794.331000
75%       3101.296400
max      13086.964800
Name: Item_Outlet_Sales, dtype: float64